File: backend/backend/Routes/TimeTable/timetable.py
from typing import Dict, Union
import datetime
from datetime import datetime as DateTime
from typing import Dict, List, Union
import re
from fastapi import APIRouter, HTTPException, Request
from utils import conn
from models import Timetable
from queries import timetable as timetable_queries
from psycopg2.errors import ForeignKeyViolation, InFailedSqlTransaction
from typing import List, Dict
from constants import default_slots
from Routes.Auth.cookie import get_user_id
import regex as re
import uuid
router = APIRouter(prefix="/schedule", tags=["schedule"])


def validate_course_schedule(data: Dict) -> Union[str, bool]:
    """
    Validates the course schedule data structure.

    Parameters:
    - data (Dict): The course schedule data containing 'courses' and 'slots'.

    Returns:
    - Union[str, bool]: Returns True if the data is valid, otherwise returns a string error message.
    """
    try:
        # Validate 'courses' section
        courses = data['courses']
        if not isinstance(courses, dict):
            return "Invalid 'courses' format. Expected a dictionary."

        for course_code, course_info in courses.items():
            if not isinstance(course_info, dict):
                return f"Invalid course info for {course_code}. Expected a dictionary."
            if "title" not in course_info or not isinstance(course_info["title"], str):
                return f"Missing or invalid 'title' for course {course_code}."

        # Validate 'slots' section
        slots = data['slots']
        if not isinstance(slots, list):
            return "Invalid 'slots' format. Expected a list."

        valid_weekdays = ['Monday', 'Tuesday', 'Wednesday',
                          'Thursday', 'Friday', 'Saturday', 'Sunday']

        # Regex pattern to match time formats like "9:00 AM" or "10:30 PM"
        time_pattern = re.compile(
            r"^(1[0-2]|0?[1-9]):([0-5]\d)\s?(AM|PM)$", re.IGNORECASE)

        for slot in slots:
            if not isinstance(slot, dict):
                return "Each slot should be a dictionary."

            # Check required fields
            required_fields = ["course_code", "day", "start_time", "end_time"]
            for field in required_fields:
                if field not in slot:
                    return f"Missing field '{field}' in slot: {slot}"

            # Validate course_code
            course_code = slot["course_code"]
            if course_code not in courses:
                return f"Invalid 'course_code' {course_code} in slot: {slot}"

            # Validate day
            day = slot["day"]
            if day not in valid_weekdays:
                return f"Invalid 'day' {day} in slot: {slot}"

            # Validate time formats
            start_time_str = slot["start_time"]
            end_time_str = slot["end_time"]

            if not time_pattern.match(start_time_str):
                return f"Invalid 'start_time' format: {start_time_str} in slot: {slot}"
            if not time_pattern.match(end_time_str):
                return f"Invalid 'end_time' format: {end_time_str} in slot: {slot}"

            # Parse times to compare
            time_format = "%I:%M %p"
            start_time = DateTime.strptime(start_time_str.upper(), time_format)
            end_time = DateTime.strptime(end_time_str.upper(), time_format)

            if start_time >= end_time:
                return f"'start_time' {start_time_str} is not earlier than 'end_time' {end_time_str} in slot: {slot}"

        # If all validations pass
        return True

    except Exception as e:
        return f"An error occurred during validation: {e}"


@router.get("/courses")
def get_timetable(request: Request) -> Timetable:
    user_id = get_user_id(request)
    try:
        query = timetable_queries.get_timetable(user_id)
        with conn.cursor() as cur:
            cur.execute(query)
            courses = cur.fetchone()
            return Timetable.from_row(courses[0])
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Internal Server Error : {e}")


@router.post("/courses")
def post_edit_timetable(request: Request, timetable: Timetable):
    user_id = get_user_id(request)
    # sanity check
    course_codes = list(timetable.courses.keys())
    # check if custom_slot_codes are not same as default slots

    validation_result = validate_course_schedule(timetable)
    if not isinstance(validation_result, bool):
        raise HTTPException(status_code=400, detail=validation_result)

    try:
        query = timetable_queries.post_timetable(user_id, timetable)
        with conn.cursor() as cur:
            cur.execute(query)
            conn.commit()
        return {"message": "Timetable Updated Successfully"}
    except Exception as e:
        conn.rollback()
        raise HTTPException(
            status_code=500, detail=f"Internal Server Error : {e}")


@router.get('/share/{code}')
def get_shared_timetable(code: str):
    try:
        query = timetable_queries.get_shared_timetable(code)
        with conn.cursor() as cur:
            cur.execute(query)
            timetable = cur.fetchone()
            if timetable is None:
                raise HTTPException(status_code=404, detail="Code Not Found")

            # convert a sql returned time string to DateTime Object\
            expiry = timetable[3]
            if expiry < DateTime.now():
                conn.commit()
                delete_query = timetable_queries.delete_shared_timetable(code)
                cur.execute(delete_query)
                conn.commit()
                raise HTTPException(
                    status_code=404, detail="Timetable has expired")

            return timetable[2]

    except HTTPException as e:
        raise e
    except Exception as e:
        conn.rollback()
        raise HTTPException(
            status_code=500, detail=f"Internal Server Error : {e}")


def generate_random_code() -> str:
    return str(uuid.uuid4().hex[:6])


@router.post('/share')
def post_share_timetable(request: Request):
    """
    Generate a unique code for the timetable, store it in db and return it
    """

    user_id = get_user_id(request)
    code = ''
    try:
        query = timetable_queries.get_timetable(user_id)
        with conn.cursor() as cur:
            cur.execute(query)
            timetable = cur.fetchone()[0]

            cur_date = DateTime.now()
            expiry_days = 120
            expiry = cur_date + datetime.timedelta(days=expiry_days)
            while True:
                code = generate_random_code()
                insert_query = timetable_queries.post_shared_timetable(
                    code, user_id, timetable, expiry)

                try:
                    cur.execute(insert_query)
                    conn.commit()
                    return {"code": code}

                except Exception as e:
                    conn.rollback()
                    continue
    except Exception as e:
        conn.rollback()
        raise HTTPException(
            status_code=500, detail=f"Internal Server Error : {e}")


@router.delete('/share/{code}')
def delete_shared_timetable(request: Request, code: str):
    user_id = get_user_id(request)
    # Check if user is the owner of this code
    try:
        query = timetable_queries.get_shared_timetable(code)
        with conn.cursor() as cur:
            cur.execute(query)
            timetable = cur.fetchone()
            if timetable is None:
                raise HTTPException(
                    status_code=404, detail="Timetable not found")

            if timetable[1] != user_id:
                raise HTTPException(
                    status_code=403, detail="You are not the owner of this timetable")

            query = timetable_queries.delete_shared_timetable(code)
            cur.execute(query)
            conn.commit()
            return {"message": "Timetable Deleted Successfully"}
    except HTTPException as e:
        raise e
    except Exception as e:
        conn.rollback()
        raise HTTPException(
            status_code=500, detail=f"Internal Server Error : {e}")

File: backend-go/internal/controller/timetable.go
package controller

import (
	"encoding/json"
	"net/http"
	"regexp"
	"strings"
	"time"

	"github.com/LambdaIITH/Dashboard/backend/config"
	timetable_queries "github.com/LambdaIITH/Dashboard/backend/internal/db"
	helpers "github.com/LambdaIITH/Dashboard/backend/internal/helpers"
	schema "github.com/LambdaIITH/Dashboard/backend/internal/schema"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

func validateCourseSchedule(data schema.Timetable) (bool, string) {
	validWeekdays := []string{"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"}
	timePattern := regexp.MustCompile(`^(1[0-2]|0?[1-9]):([0-5]\d)\s?(AM|PM)$`)

	// Validate 'courses'
	for courseCode, courseInfo := range data.Courses {
		if strings.TrimSpace(courseCode) == "" {
			return false, "Course code cannot be empty."
		}
		if courseInfo["title"] == "" {
			return false, "Course title cannot be empty."
		}
	}

	// Validate 'slots'
	for _, slot := range data.Slots {
		requiredFields := []string{"course_code", "day", "start_time", "end_time"}
		for _, field := range requiredFields {
			if slot[field] == "" {
				return false, "Missing field '" + field + "' in slot."
			}
		}

		if !helpers.Contains(validWeekdays, slot["day"]) {
			return false, "Invalid day: " + slot["day"]
		}

		if !helpers.MatchRegex(slot["start_time"], timePattern) || !helpers.MatchRegex(slot["end_time"], timePattern) {
			return false, "Invalid time format in slot."
		}

		startTime, err := time.Parse("3:04 PM", slot["start_time"])
		if err != nil {
			return false, "Invalid start_time format."
		}
		endTime, err := time.Parse("3:04 PM", slot["end_time"])
		if err != nil {
			return false, "Invalid end_time format."
		}
		if !startTime.Before(endTime) {
			return false, "start_time must be earlier than end_time."
		}
	}

	return true, ""
}

func GetTimetable(c *gin.Context) {
	userID, err := helpers.GetUserID(c.Request)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
		return
	}

	timetableJSON, err := timetable_queries.GetTimetable(c, userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch timetable."})
		return
	}
	var timetable schema.Timetable
	if err := json.Unmarshal([]byte(timetableJSON), &timetable); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse timetable."})
		return
	}

	c.JSON(http.StatusOK, timetable)
}

func PostEditTimetable(c *gin.Context) {
	var timetable schema.Timetable
	userID, err := helpers.GetUserID(c.Request)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
		return
	}

	if err := c.ShouldBindJSON(&timetable); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	valid, errMsg := validateCourseSchedule(timetable)
	if !valid {
		c.JSON(http.StatusBadRequest, gin.H{"error": errMsg})
		return
	}

	_, err = timetable_queries.PostTimetable(c, userID, timetable)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update timetable."})
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"message": "Timetable successfully updated",
	})

}

func GetSharedTimetable(c *gin.Context) {
	code := c.Param("code")

	if code == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Code is required"})
		return
	}

	timetableJSON, err := timetable_queries.GetSharedTimetable(c, code)
	if err != nil {

		if err.Error() == "timetable has expired" {
			c.JSON(http.StatusNotFound, gin.H{"error": "Timetable has expired"})
		} else {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		}
		return
	}
	var timetable schema.Timetable
	if err := json.Unmarshal([]byte(timetableJSON), &timetable); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse timetable."})
		return
	}

	c.JSON(http.StatusOK, timetable)
}

func GenerateRandomCode() string {
	uuidValue := uuid.New()
	code := uuidValue.String()[:6]
	return strings.ToUpper(code)
}

func PostSharedTimetable(c *gin.Context) {
	userID, err := helpers.GetUserID(c.Request)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
		return
	}
	var timetable schema.Timetable
	if err := c.ShouldBindJSON(&timetable); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid JSON format"})
		return
	}
	code := GenerateRandomCode()

	valid, errMsg := validateCourseSchedule(timetable)
	if !valid {
		c.JSON(http.StatusBadRequest, gin.H{"error": errMsg})
		return
	}

	curDate := time.Now()
	expiryDays := 120
	expiry := curDate.AddDate(0, 0, expiryDays)
	result, err := timetable_queries.PostSharedTimetable(c, code, userID, timetable, expiry)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to share timetable."})
		return
	}

	c.JSON(http.StatusOK, result)
}

func DeleteSharedTimetable(c *gin.Context) {
	user_id, err := helpers.GetUserID(c.Request)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "User not found"})
	}
	code := c.Param("code")
	if code == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Code is required"})
		return
	}
	query := "SELECT timetable FROM shared_timetable WHERE code = $1"
	var id int
	err1 := config.DB.QueryRow(c, query, code).Scan(&id)
	if err1 != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Time Table not found."})
		return
	}
	if id != user_id {
		c.JSON(http.StatusBadRequest, gin.H{"error": "You are not the owner of this timetable"})
	}

	result, err := timetable_queries.DeleteSharedTimetable(c, code)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete timetable."})
		return
	}

	c.JSON(http.StatusOK, result)
}
